File: frontend/src/models/items/Event.ts
import Item from '@/models/Items'
import { LanguageCode } from '@/models/Language'

/**
 * @interface Event
 * @extends Item
 * @description Describes an event from the backend and API responses
 *
 * @property {string} location - The location of the event
 * @property {string} start_date - The start date of the event
 * @property {string} end_date - The end date of the event
 * @property {number} duration - The duration of the event
 * @property {string} background_color - The background color of the event
 * @property {EventTranslation[]} translations - The translations of the event
 */
export default interface Event extends Item {
  event_id: string
  location: string
  start_date: string
  duration: number
  background_color: string
  translations: EventTranslation[]
}

/**
 * @interface EventTranslation
 * @description Holds the translations of each event, is at least 1:1 with the Event model and is used to display the event in different languages
 *
 * @property {string} title - The title of the event
 * @property {string} description - The description of the event (optional)
 * @property {string} main_image_url - The main image URL of the event (optional)
 * @property {string[]} sub_image_urls - The sub image URLs of the event (optional)
 * @property {LanguageCode} language_code - The language code of the event
 */
export interface EventTranslation {
  title: string
  description?: string
  main_image_url?: string
  sub_image_urls?: string[]
  language_code: LanguageCode
}

File: apps/backend/core-api/dto/content/event.py
import uuid

from pydantic import BaseModel

from dto.content.item import ItemDTO


class EventTranslationDTO(BaseModel):
    title: str
    description: str
    main_image_url: str | None = None
    sub_image_urls: list[str] | None = []
    language_code: str


class EventDTO(ItemDTO):
    event_id: uuid.UUID
    start_date: str
    duration: int
    background_color: str
    location: str
    is_inherited: bool
    translations: list[EventTranslationDTO] | None = []

    @classmethod
    def from_orm_with_language(cls, obj, language_code: str) -> "EventDTO":
        """
        Create a EventDTO from an ORM object and a language code.
        """
        # Filter translations by the specified language code
        translations = [
            translation
            for translation in obj.translations
            if translation.language_code == language_code
        ]

        # Create a new EventDTO instance with the filtered translations
        return cls(
            event_id=obj.event_id,
            start_date=obj.start_date,
            duration=obj.duration,
            background_color=obj.background_color,
            location=obj.location,
            is_inherited=obj.is_inherited,
            translations=translations,
        )
